test/functional/Valhalla/src_lw5/org/openj9/test/lworld/NullRestrictedTypeOptTests.java

```diff
@@ -0,0 +1,424 @@
+		FlattenablePair! p4;
+
+		if (iter % 2 == 0) {
+			p3 = p2;
+			p4 = p1;
+		} else {
+			p3 = p1;
+			p4 = p2;
+		}
+		result += p3.x + p4.y;
+	}
+
+	static private void evalTestEA1b() {
+		for (int j = 0; j < 100; j++) {
+			// Test situation where EA could apply to value p1,
+			// but might have to allocate contiguously
+			FlattenablePair! p1 = new FlattenablePair(1, 2);
+			FlattenablePair! p2 = parr[0];
+
+			FlattenablePair! p3;
+			FlattenablePair! p4;
+
+			if (j % 2 == 0) {
+				p3 = p2;
+				p4 = p1;
+			} else {
+				p3 = p1;
+				p4 = p2;
+			}
+			result += p3.x + p4.y;
+		}
+	}
+
+	public static FlattenablePair! testEA2Field = new FlattenablePair(0,0);
+
+	@Test(priority=1)
+	static public void testEAOpportunity2() throws Throwable {
+		for (int i = 0; i < 10000; i++) {
+			result = 0;
+			evalTestEA2(-1);  // No escape
+			assertEquals(result, 200);
+			assertEquals(testEA2Field, new FlattenablePair(0,0));
+		}
+		evalTestEA2(99);  // Escape for index 99
+		assertEquals(testEA2Field, new FlattenablePair(2,1));
+		evalTestEA2(98);  // Escape for index 98
+		assertEquals(testEA2Field, new FlattenablePair(1,2));
+	}
+
+	static private void evalTestEA2(int escapePoint) {
+		int x = 1; int y = 2;
+		int[] nextVal = {0, 2, 1};
+
+		for (int i = 0; i < 100; i++) {
+			FlattenablePair! p1 = new FlattenablePair(x, y);
+			int updatex = nextVal[x];
+			int updatey = nextVal[y];
+			x = updatex;
+			y = updatey;
+			if (p1.x*p1.y != 2 || escapePoint == i) testEA2Field = p1;  // Value might escape
+
+			result += x*y;
+		}
+	}
+
+	@Test(priority=1)
+	static public void testEAOpportunity3() throws Throwable {
+		for (int i = 0; i < 1000; i++) {
+			result = 0;
+			evalTestEA3();
+			assertEquals(result, 1000);
+		}
+	}
+
+	static private void evalTestEA3() {
+		for (int i = 0; i < 100; i++) {
+			// Test potential stack allocation of array of value type
+			FlattenablePair![] arr = new FlattenablePair![2];
+			arr[0] = new FlattenablePair(1, 2);
+			arr[1] = new FlattenablePair(3, 4);
+			for (int j = 0; j < arr.length; j++) {
+				FlattenablePair! val = arr[j%arr.length];
+				result += val.x + val.y;
+			}
+		}
+	}
+
+	/* A value class with an implicit constructor (eligible to be null restricted)
+	 * with null restricted value class fields.
+	 */
+	public value static class NestedFlattenablePair {
+		public final FlattenablePair! p1;
+		public final FlattenablePair! p2;
+
+		public implicit NestedFlattenablePair();
+
+		public NestedFlattenablePair(int i, int j, int m, int n) {
+			this.p1 = new FlattenablePair(i, j);
+			this.p2 = new FlattenablePair(m, n);
+		}
+
+		public NestedFlattenablePair(FlattenablePair p1, FlattenablePair p2) {
+			this.p1 = new FlattenablePair(p1);
+			this.p2 = new FlattenablePair(p2);
+		}
+	}
+
+	// An array whose component type is a null-restricted value class
+	public static NestedFlattenablePair![] nestedflattenablearr = new NestedFlattenablePair[] { new NestedFlattenablePair(11, 12, 13, 14) };
```

Comment:
Sorry - I missed this one the first time around.  I think the right-hand side of this one needs to allocate a null-restricted array as well.

Reply:
woops thanks for that. I've updated this array.